flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java

```diff
@@ -516,7 +525,14 @@ public void handleOperatorEvent(OperatorEvent event) {
             checkWatermarkAlignment();
         } else if (event instanceof AddSplitEvent) {
             try {
-                sourceReader.addSplits(((AddSplitEvent<SplitT>) event).splits(splitSerializer));
+                List<SplitT> newSplits = ((AddSplitEvent<SplitT>) event).splits(splitSerializer);
+                if (operatingMode == OperatingMode.OUTPUT_NOT_INITIALIZED) {
+                    outputPendingSplits.addAll(newSplits);
+                } else {
+                    newSplits.forEach(
```

Comment:
So I guess we have two cases:
- bulk addition of initial splits before first `emitNext`
- bulk addition of initial splits after first `emitNext`

You should ensure that both cases are covered by tests. I'd also recommend to add inline (java)docs that spell out the cases explicitly. For that I'd probably extract this `forEach` to a private method (e.g., `setupInitialSplit` or find a better name).